### PR TITLE
fix(ppo): reject non-finite sequence-mask importance weights

### LINKED ISSUE #1295: [Bug] seq-mask-tis skips token-level truncation and can produce non-finite PPO updates
## Summary

The `seq-mask-tis` branch says that its sequence-level geometric mean is used only for filtering and that correction coefficients still use token-level truncated importance sampling (TIS). However, it exponentiates `old_log_probs - rollout_log_probs` without applying the configured low/high clamp.

This makes the mode differ from its documented semantics even for finite ratios. With a sufficiently stale rollout, `exp()` overflows; an accepted sequence then has an infinite loss/gradient, while a rejected sequence can evaluate `0 * inf` and produce `NaN`.

**Severity: medium-high.** The mode is opt-in, but once selected a single extreme off-policy sequence can poison an expensive PPO optimizer step and subsequent training.

## Affected code

At commit `bc71bb19464aca306b33080b2d2bb45d154e2f49`, `openrlhf/models/loss.py` contains:

```python
elif self.vllm_is_correction_type == "seq-mask-tis":
    # seq-mask-tis: use sequence-level geometric mean only for filtering,
    # correction coefficients still use TIS (token-level clamp)
    seq_log_ratio = masked_mean(rollout_log_ratio, action_mask, dim=-1)
    seq_is = torch.exp(seq_log_ratio)
    seq_mask = (seq_is >= low_threshold) & (seq_is <= high_threshold)
    vllm_is = torch.exp(rollout_log_ratio).detach()
    loss = seq_mask.unsqueeze(-1) * vllm_is * loss
```

The regular `tis` branch immediately below correctly applies `.clamp(min=low_threshold, max=high_threshold)`.

## Reproduction

CPU-only, using PyTorch 2.13.0. All log-probabilities below are non-positive.

```python
import math
import torch
from openrlhf.models.loss import PolicyLoss

loss_fn = PolicyLoss(
    policy_loss_type="ppo",
    enable_vllm_is_correction=True,
    vllm_is_truncated_threshold=[0.5, 5.0],
    vllm_is_correction_type="seq-mask-tis",
)

def run(rollout_log_probs):
    old_log_probs = torch.tensor([[0.0, -1000.0]])
    log_probs = old_log_probs.clone().requires_grad_(True)
    loss, *_ = loss_fn(
        log_probs,
        old_log_probs,
        torch.ones_like(log_probs),
        action_mask=torch.ones_like(log_probs),
        rollout_log_probs=torch.tensor([rollout_log_probs]),
    )
    loss.backward()
    print(loss, log_probs.grad, torch.isfinite(loss), torch.isfinite(log_probs.grad).all())

# Finite token ratios [10, 0.1] have geometric mean 1 and pass the sequence filter.
d = math.log(10.0)
old_log_probs = torch.tensor([[-0.1, -0.1 - d]])
log_probs = old_log_probs.clone().requires_grad_(True)
moderate_loss, *_ = loss_fn(
    log_probs,
    old_log_probs,
    torch.ones_like(log_probs),
    action_mask=torch.ones_like(log_probs),
    rollout_log_probs=torch.tensor([[-0.1 - d, -0.1]]),
)
print(moderate_loss)  # -5.05 on main; TIS-clamped result should be -2.75

# Token ratios are exp([1000, -1000]); geometric mean is 1, so the sequence is accepted.
run([-1000.0, 0.0])

# Token ratios are exp([1000, -996]); geometric mean is exp(2) > 5, so it is rejected.
run([-1000.0, -4.0])
```

Actual result on `main`:

```text
moderate_loss=-5.05 (configured [0.5, 5.0] bounds were bypassed)
loss=-inf, gradient contains -inf, finite=False
loss=nan, gradient contains nan, finite=False
```

The rejected case becomes `False * inf`, which is `NaN` rather than zero.

## Expected behavior

After sequence-level filtering, token coefficients should be clamped to `[low_threshold, high_threshold]`, exactly as the comment and the `tis` implementation specify. With the thresholds above, the accepted case should use weights `[5.0, 0.5]` and produce a finite loss of `-2.75`; the rejected case should produce finite zero loss/gradients.

## Suggested fix

Apply the same clamp as the `tis` branch before detaching:

```python
vllm_is = torch.exp(rollout_log_ratio).clamp(min=low_threshold, max=high_threshold).detach()
```

Add regression coverage for both an accepted extreme sequence and a rejected extreme sequence, asserting finite loss and gradients.

## Reachability

This is reachable through the normal CLI path:

```text
--algo.advantage.is_correction_enable
--algo.advantage.is_correction_type seq-mask-tis
  -> vLLM rollout_log_probs
  -> Experience.rollout_log_probs
  -> PolicyLoss(..., policy_loss_type="ppo")
```

## Duplicate check

I searched open and closed issues and all PRs for `seq-mask-tis`, `token clamp`, `TIS overflow`, and `importance sampling NaN`; no existing report or fix covers this branch.

PR #1293 is adjacent but distinct: it bounds the exponent in the `gspo` policy-ratio branch (`loss.py:170-178`). This bug is in PPO's `seq-mask-tis` correction branch (`loss.py:207-214`) and still reproduces with #1293 applied. PR #1240 addresses invalid values at masked positions; this reproduction uses an all-ones action mask.

### PR DESCRIPTION
**Pull request by ai-yang**
## Summary

Prevent an accepted sequence from injecting non-finite importance weights into the PPO loss or gradients. Closes #1295.

After the IS-correction refactor in #1339, `level=seq, mode=mask` intentionally retains raw per-token importance weights for accepted sequences. This PR preserves that behavior for finite weights; it does not restore the old token-level TIS clamp.

The sequence gate alone is insufficient: log ratios `[1000, -1000]` have a geometric-mean ratio of 1, but their token weights are `[inf, 0]`. The aggregate gate passes and the loss becomes non-finite.

The fix rejects a sequence if any active token importance weight is non-finite, and zeroes non-finite padding weights before multiplication. Ordinary finite weights, token-level modes, and the selected aggregate statistic are unchanged. The existing filter metric includes sequences rejected by this numerical check. This deliberately proposes rejection of unrepresentable sequence weights rather than introducing a new finite clipping threshold; reviewer feedback on that policy is welcome.

## Validation

Validated against main `dc2a7ad` on 2026-09-21, using real CPU PyTorch 2.13.0:

- Added 13 cases covering ratio, binary-KL and TV sequence gates; float16, bfloat16, float32 and float64; unaffected batch members; padding; loss, gradient and filter-metric results.
- On unpatched main, the new cases produce **11 failed, 2 passed**. The two already-passing cases are low-precision binary-KL gates that reject the example before weighting.
- With the fix, the complete CPU suite reports **216 passed, 2 skipped, 14 warnings**. Optional dependency tests are skipped; the suite uses the existing import-only DeepSpeed shim, not a distributed training run.
- Black, isort, `compileall` and `git diff --check` pass.
- `bash -n` passes for 17 example scripts. The unchanged upstream SLURM template contains the literal placeholder `OPENRLHF_PATH=<OPENRLHF_ROOT_PATH>` and fails parsing until configured; no shell scripts are changed by this PR.

The production diff against current main is six added lines in `PolicyLoss`; the remaining changes are regressions. The branch incorporates main without rewriting its existing history. No GPU rollout or end-to-end training result is claimed.

OpenAI Codex assisted with conflict resolution, implementation, and validation.

### CODE REVIEW

**Review by Excelius-Wang — COMMENTED** [review]
The sequence filter now stays separate from token-level truncation, which matches the intended seq-mask-tis behavior.

I checked a sequence whose geometric mean ratio is 1 but whose token ratios are 10 and 0.1: the weights are clipped to 5 and 0.5, including in the backward pass. The extreme rejected-sequence case also gives zero loss and zero, finite gradients. These checks fail with the old branch and pass here. My validation was on CPU; I didn't run a rollout or training job.
